### PR TITLE
Merge Protocol Defense Tasks into Dev

### PR DESCRIPTION
**Pull request by koloz193**
# What ❔

The purpose of these tasks is to make the protocol more resilient while increasing the quality of life when it comes to development, interop, and having a correct running system.

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.

### CODE REVIEW

File: l1-contracts/contracts/state-transition/chain-deps/facets/Executor.sol [hunk 1]
```diff
@@ -53,7 +56,12 @@ contract ExecutorFacet is ZkSyncHyperchainBase, IExecutor {
         bytes32[] memory blobCommitments = new bytes32[](MAX_NUMBER_OF_BLOBS);
         if (pricingMode == PubdataPricingMode.Validium) {
             // skipping data validation for validium, we just check that the data is empty
-            require(_newBatch.pubdataCommitments.length == 1, "EF: v0l");
+            if (logOutput.pubdataHash != 0x00) {
```
Comment (vladbochok) [inline l1-contracts/contracts/state-transition/chain-deps/facets/Executor.sol:59]: This is new error, intended @koloz193 ?

File: .codespell/wordlist.txt [hunk 2]
```diff
@@ -4,3 +4,4 @@ ue
 tha
 bu
 crate
+DNE
```
Comment (StanislavBreadless) [inline .codespell/wordlist.txt:7]: what is it?

File: l1-contracts/contracts/bridge/L1SharedBridge.sol [hunk 3]
```diff
@@ -137,21 +141,27 @@ contract L1SharedBridge is IL1SharedBridge, ReentrancyGuard, Ownable2StepUpgrade
...
+            revert SharedBridgeValueAlreadySet(SharedBridgeKey.PostUpgradeFirstBatch);
+        }
         eraPostDiamondUpgradeFirstBatch = _eraPostDiamondUpgradeFirstBatch;
     }
 
     /// @dev This sets the first post upgrade batch for era, used to check old token withdrawals
     /// @param _eraPostLegacyBridgeUpgradeFirstBatch The first batch number on the zkSync Era Diamond Proxy that was settled after legacy bridge upgrade.
     function setEraPostLegacyBridgeUpgradeFirstBatch(uint256 _eraPostLegacyBridgeUpgradeFirstBatch) external onlyOwner {
-        require(eraPostLegacyBridgeUpgradeFirstBatch == 0, "ShB: eFPUB already set");
+        if (eraPostLegacyBridgeUpgradeFirstBatch != 0) {
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/L1SharedBridge.sol:162]: Let's delete these methods altogether cc @kelemeno

File: l1-contracts/contracts/bridge/L1SharedBridge.sol [hunk 4]
```diff
@@ -264,27 +293,43 @@ contract L1SharedBridge is IL1SharedBridge, ReentrancyGuard, Ownable2StepUpgrade
...
+            if (msg.value != 0) {
+                revert NonEmptyMsgValue();
+            }
             amount = _depositAmount;
 
             uint256 withdrawAmount = _depositFunds(_prevMsgSender, IERC20(_l1Token), _depositAmount);
-            require(withdrawAmount == _depositAmount, "5T"); // The token has non-standard transfer logic
+            // The token has non-standard transfer logic
+            if (withdrawAmount != _depositAmount) {
+                revert WithdrawIncorrectAmount();
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/L1SharedBridge.sol:326]: In similar places the arguments were provided in the error:

```
                revert ValueMismatch(_amount, amount);
```
Consider providing them here as well

File: l1-contracts/contracts/bridge/L1SharedBridge.sol [hunk 5]
```diff
@@ -432,14 +487,18 @@ contract L1SharedBridge is IL1SharedBridge, ReentrancyGuard, Ownable2StepUpgrade
             if (notCheckedInLegacyBridgeOrWeCanCheckDeposit) {
                 bytes32 dataHash = depositHappened[_chainId][_l2TxHash];
                 bytes32 txDataHash = keccak256(abi.encode(_depositSender, _l1Token, _amount));
-                require(dataHash == txDataHash, "ShB: d.it not hap");
+                if (dataHash != txDataHash) {
+                    revert DepositDNE();
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/L1SharedBridge.sol:491]: what is DNE?

File: l1-contracts/contracts/state-transition/chain-deps/facets/Executor.sol (hunk 1 above)
Comment (koloz193) [reply l1-contracts/contracts/state-transition/chain-deps/facets/Executor.sol:59]: let me remove it, i think it was present before we merged in release-v23

File: .codespell/wordlist.txt (hunk 2 above)
Comment (koloz193) [reply .codespell/wordlist.txt:7]: does not exist but i can update it to be more explicit

File: l2-contracts/contracts/bridge/L2StandardERC20.sol [hunk 6]
```diff
@@ -167,19 +167,22 @@ contract L2StandardERC20 is ERC20PermitUpgradeable, IL2StandardToken, ERC1967Upg
 
     function name() public view override returns (string memory) {
         // If method is not available, behave like a token that does not implement this method - revert on call.
-        if (availableGetters.ignoreName) revert Unimplemented();
+        // solhint-disable-next-line reason-string, gas-custom-errors
+        if (availableGetters.ignoreName) revert();
```
[the remark this answers is not in the document]
Comment (koloz193) [reply l2-contracts/contracts/bridge/L2StandardERC20.sol:171]: @vladbochok had mentioned we want the same behavior as if the method didn't exist

File: system-contracts/bootloader/bootloader.yul [hunk 7]
```diff
@@ -12,13 +12,13 @@ object "Bootloader" {
             // While we definitely cannot control the pubdata price on L1,
             // we need to check the operator does not provide any absurd numbers there
             function MAX_ALLOWED_FAIR_PUBDATA_PRICE() -> ret {
-                // 1M gwei
-                ret := 1000000000000000
+                // 2^64 - 1 gwei
+                ret := 18446744073709551615
```
[the remark this answers is not in the document]
Comment (koloz193) [reply system-contracts/bootloader/bootloader.yul:16]: i just tested it and they seem to be sufficient

**Review by StanislavBreadless — APPROVED** [review]

File: l1-contracts/scripts/setup-legacy-bridge-era.ts [hunk 8]
```diff
@@ -15,8 +15,8 @@ import { web3Provider, GAS_MULTIPLIER } from "./utils";
 import { deployedAddressesFromEnv } from "../src.ts/deploy-utils";
 import { ethTestConfig, getAddressFromEnv } from "../src.ts/utils";
 import { hashL2Bytecode } from "../../l2-contracts/src/utils";
-import { Provider } from "zksync-ethers";
-import beaconProxy = require("../../l2-contracts/artifacts-zk/@openzeppelin/contracts/proxy/beacon/BeaconProxy.sol/BeaconProxy.json");
+import { Provider } from "zksync-web3";
```
Comment (StanislavBreadless) [inline l1-contracts/scripts/setup-legacy-bridge-era.ts:18]: Why do we use this package? It is somewhat outdated compared to zksync-ethers

File: gas-bound-caller/contracts/GasBoundCallerErrors.sol [hunk 9]
```diff
@@ -0,0 +1,8 @@
+// SPDX-License-Identifier: MIT
+// We use a floating point pragma here so it can be used within other projects that interact with the zkSync ecosystem without using our exact pragma version.
```
[the remark this answers is not in the document]
Comment (koloz193) [reply gas-bound-caller/contracts/GasBoundCallerErrors.sol:2]: use the comment across all contracts?

**Review by StanislavBreadless — APPROVED** [review]

File: gas-bound-caller/contracts/GasBoundCallerErrors.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (vladbochok) [reply gas-bound-caller/contracts/GasBoundCallerErrors.sol]: Use floting pragma on all contracts, how do we know what contract devs will import?

File: l1-contracts/test/foundry/integration/deploy-scripts/script-out/output-deploy-l1.toml [hunk 10]
```diff
@@ -0,0 +1,55 @@
+create2_factory_addr = "0x4e59b44847b379578588920cA78FbF26c0B4956C"
```
Comment (vladbochok) [inline l1-contracts/test/foundry/integration/deploy-scripts/script-out/output-deploy-l1.toml:1]: Artifacts?

File: system-contracts/contracts/L1Messenger.sol [hunk 11]
```diff
@@ -209,7 +209,7 @@ contract L1Messenger is IL1Messenger, ISystemContract {
         calldataPtr += 4;
 
         bytes32[] memory l2ToL1LogsTreeArray = new bytes32[](L2_TO_L1_LOGS_MERKLE_TREE_LEAVES);
-        bytes32 reconstructedChainedLogsHash;
+        bytes32 reconstructedChainedLogsHash = bytes32(0);
```
Comment (vladbochok) [inline system-contracts/contracts/L1Messenger.sol:212]: Why?

File: system-contracts/contracts/L1Messenger.sol [hunk 12]
```diff
@@ -238,7 +238,7 @@ contract L1Messenger is IL1Messenger, ISystemContract {
         /// Check messages
         uint32 numberOfMessages = uint32(bytes4(_totalL2ToL1PubdataAndStateDiffs[calldataPtr:calldataPtr + 4]));
         calldataPtr += 4;
-        bytes32 reconstructedChainedMessagesHash;
+        bytes32 reconstructedChainedMessagesHash = bytes32(0);
```
Comment (vladbochok) [inline system-contracts/contracts/L1Messenger.sol:241]: Why?

File: system-contracts/contracts/L1Messenger.sol [hunk 13]
```diff
@@ -255,7 +255,7 @@ contract L1Messenger is IL1Messenger, ISystemContract {
         /// Check bytecodes
         uint32 numberOfBytecodes = uint32(bytes4(_totalL2ToL1PubdataAndStateDiffs[calldataPtr:calldataPtr + 4]));
         calldataPtr += 4;
-        bytes32 reconstructedChainedL1BytecodesRevealDataHash;
+        bytes32 reconstructedChainedL1BytecodesRevealDataHash = bytes32(0);
```
Comment (vladbochok) [inline system-contracts/contracts/L1Messenger.sol:258]: Why?

**Review by vladbochok — COMMENTED** [review]
Please fix merge conflicts

File: l1-contracts/contracts/state-transition/chain-deps/facets/Executor.sol [hunk 14]
```diff
@@ -147,7 +169,7 @@ contract ExecutorFacet is ZkSyncHyperchainBase, IExecutor {
 
         // Used as bitmap to set/check log processing happens exactly once.
         // See SystemLogKey enum in Constants.sol for ordering.
-        uint256 processedLogs;
+        uint256 processedLogs = 0;
```
[the remark this answers is not in the document]
Comment (koloz193) [reply l1-contracts/contracts/state-transition/chain-deps/facets/Executor.sol:172]: for all these OZ suggested adding in the default values. i dont have a strong opinion on keeping or removing

File: l1-contracts/contracts/state-transition/chain-interfaces/IMailbox.sol [hunk 15]
```diff
@@ -121,8 +121,8 @@ interface IMailbox is IZkSyncHyperchainBase {
     /// @param factoryDeps An array of bytecodes that were shown in the L1 public data.
     /// Will be marked as known bytecodes in L2
     event NewPriorityRequest(
-        uint256 txId,
-        bytes32 txHash,
+        uint256 indexed txId,
+        bytes32 indexed txHash,
```
[the remark this answers is not in the document]
Comment (koloz193) [reply l1-contracts/contracts/state-transition/chain-interfaces/IMailbox.sol:125]: i will double check

File: l1-contracts/contracts/dev-contracts/test/DummySharedBridge.sol [hunk 16]
```diff
@@ -2,14 +2,14 @@
...
 
-import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {IERC20} from "@openzeppelin/contracts-v4/token/ERC20/IERC20.sol";
 
 import {L2TransactionRequestTwoBridgesInner} from "../../bridgehub/IBridgehub.sol";
 import {TWO_BRIDGES_MAGIC_VALUE, ETH_TOKEN_ADDRESS} from "../../common/Config.sol";
-import {PausableUpgradeable} from "@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol";
+import {PausableUpgradeable} from "@openzeppelin/contracts-upgradeable-v4/security/PausableUpgradeable.sol";
 import {Ownable2StepUpgradeable} from "@openzeppelin/contracts-upgradeable/access/Ownable2StepUpgradeable.sol";
 import {UnsafeBytes} from "contracts/common/libraries/UnsafeBytes.sol";
```
Comment (benceharomi) [inline l1-contracts/contracts/dev-contracts/test/DummySharedBridge.sol:11]:
```suggestion
import {UnsafeBytes} from "../../common/libraries/UnsafeBytes.sol";
```

Comment (benceharomi) [reply l1-contracts/contracts/dev-contracts/test/DummySharedBridge.sol:11]: Relative import is used everywhere else. I know it is not introduced with this PR, but please include this change too :)

Comment (koloz193) [reply l1-contracts/contracts/dev-contracts/test/DummySharedBridge.sol:11]: i dont believe its used we can probably just remove it

File: system-contracts/contracts/L1Messenger.sol (hunk 11 above)
Comment (koloz193) [reply system-contracts/contracts/L1Messenger.sol:212]: for all these OZ suggested adding in the default values. i dont have a strong opinion on keeping or removing

File: system-contracts/contracts/L1Messenger.sol (hunk 12 above)
Comment (koloz193) [reply system-contracts/contracts/L1Messenger.sol:241]: for all these OZ suggested adding in the default values. i dont have a strong opinion on keeping or removing

File: system-contracts/contracts/L1Messenger.sol (hunk 13 above)
Comment (koloz193) [reply system-contracts/contracts/L1Messenger.sol:258]: for all these OZ suggested adding in the default values. i dont have a strong opinion on keeping or removing

File: l1-contracts/contracts/state-transition/StateTransitionManager.sol [hunk 17]
```diff
@@ -104,7 +107,7 @@ contract StateTransitionManager is IStateTransitionManager, ReentrancyGuard, Own
         chainAddresses = new address[](keys.length);
         uint256 keysLength = keys.length;
         for (uint256 i = 0; i < keysLength; ++i) {
-            chainAddresses[i] = hyperchainMap.get(i);
+            chainAddresses[i] = hyperchainMap.get(keys[i]);
```
[the remark this answers is not in the document]
Comment (StanislavBreadless) [reply l1-contracts/contracts/state-transition/StateTransitionManager.sol:110]: @vladbochok I did it a few days ago

[the remark this answers is not in the document]
Comment (StanislavBreadless) [reply l1-contracts/contracts/state-transition/StateTransitionManager.sol:110]: And even sooner it was done by @Deniallugo : https://github.com/matter-labs/era-contracts/pull/696

**Review by vladbochok — APPROVED** [review]

File: l1-contracts/deploy-scripts/DecentralizeGovernanceUpgradeScript.s.sol [hunk 18]
```diff
@@ -1,4 +1,5 @@
 // SPDX-License-Identifier: MIT
```
Comment (benceharomi) [inline l1-contracts/deploy-scripts/DecentralizeGovernanceUpgradeScript.s.sol:1]: `ProxyAdmin` and `ITransparentUpgradeableProxy` imports are not updated to use the v4 alias:
```
import {ProxyAdmin} from "@openzeppelin/contracts/proxy/transparent/ProxyAdmin.sol";
import {ITransparentUpgradeableProxy} from "@openzeppelin/contracts/proxy/transparent/TransparentUpgradeableProxy.sol";
```
It should be changed to:
```
import {ProxyAdmin} from "@openzeppelin/contracts-v4/proxy/transparent/ProxyAdmin.sol";
import {ITransparentUpgradeableProxy} from "@openzeppelin/contracts-v4/proxy/transparent/TransparentUpgradeableProxy.sol";
```

**Comment by benceharomi** [comment]
Nothing is depending on the OZ v5 contracts packages yet, I would remove them from `lib` and `package.json` until they are actually needed.

**Pull request merged**